pkg/backend/filestate/backend.go

```diff
@@ -241,6 +240,14 @@ func Login(ctx context.Context, d diag.Sink, url string, project *workspace.Proj
 	return be, workspace.StoreAccount(be.URL(), workspace.Account{}, true)
 }
 
+func (b *localBackend) getReference(ref backend.StackReference) (*localBackendReference, error) {
+	localStackRef, is := ref.(*localBackendReference)
+	if !is {
+		return nil, fmt.Errorf("bad stack reference type")
+	}
+	return localStackRef, nil
```

Comment:
I wonder if this should actually be more strict. In the project changes I added a `*localBackend` field to `localBackendReference` and maybe this should check that the references isn't just the right type, but also comes from the backend object that is now trying to use it.

Reply:
I may have removed the localBackend field. 😆 

I suspect the type check here is enough: outside callers cannot construct their own localBackendReference.
I can't imagine many likely scenarios where we'll need to guard against two filestate backends passing bad references to each other. However, if that changes, we can guard against it either in the way you suggest or with a cheap numeric identifier assigned to the localBackend and passed to the localBackendReference.
(I think keeping localBackendReference as a data-only object is desirable until we add meaningfully complex behavior to it.)

Reply:
Yeh was just a thought, but given we'll need to re-add `*localBackend` to do the project name lookup in ParseStackReference it would be very doable.

Reply:
Sorry, I don't understand. Why do we need to add localBackend to localBackendReference?
ParseStackReference is a method on localBackend, so it can do the lookup as needed.

For context, this is the change I made to drop the dependency:
https://github.com/pulumi/pulumi/pull/12134/commits/980b3afe4394acbfdede065b43bb4e4a8d45ef7f
localBackendReference held onto localBackend only to look up the current project, so it could shorten the `String()` representation. I figured it can do that with the name of the current project at creation time instead.
(Accessing localBackend.currentProject on String has implications to thread-safety: there's no guarantee that String is not called concurrently; so if we end up putting the reference back. localBackend.currentProject will probably become an atomic.Pointer.)

Reply:
> Why do we need to add localBackend to localBackendReference?

To match the behavior of httpBackend. You'll change the output of `pulumi new` if you don't use the current project because we parse the stack reference first then make the project and set it with "SetCurrentProject".

It might be ok to change this, but filestate and httpstate should be the same in regards to this.

Reply:
Ack. I'll add the behavior back in the filestate PR.
(This PR will be unaffected by that.)

Reply:
Added back in https://github.com/pulumi/pulumi/pull/12134/commits/e3e40b1987b689f24384bb784c4e393230a2defe and fixed race in https://github.com/pulumi/pulumi/pull/12134/commits/c91e4930c23edbe4b1faaeef96bea3ad9fc68537.
(This is just FYI for changes to the draft PR. Real PRs incoming.)